src/VisualStudio/CSharp/Impl/LanguageService/CSharpCreateServicesOnTextViewConnection.cs

```diff
@@ -29,5 +39,43 @@ internal class CSharpCreateServicesOnTextViewConnection : AbstractCreateServices
             : base(workspace, listenerProvider, threadingContext, LanguageNames.CSharp)
         {
         }
+
+        protected override Task InitializeServiceForOpenedDocumentAsync(Document document)
+        {
+            // Only pre-populate cache if import completion is enabled
+            if (this.Workspace.Options.GetOption(CompletionOptions.ShowItemsFromUnimportedNamespaces, LanguageNames.CSharp) != true)
+                return Task.CompletedTask;
+
+            var documentId = document.Id;
+            lock (_object)
+            {
+                _unimportedTypeTask = _unimportedTypeTask.ContinueWith(async _
+                    => await PopulateTypeImportCompletionCacheAsync(this.Workspace.CurrentSolution.GetDocument(documentId)).ConfigureAwait(false), TaskScheduler.Default);
+
+                _unimportedExtensionMethodTask = _unimportedExtensionMethodTask.ContinueWith(async _
+                    => await PopulateExtensionMethodImportCompletionCacheAsync(this.Workspace.CurrentSolution.GetDocument(documentId)).ConfigureAwait(false), TaskScheduler.Default);
```

Comment:
no need for async/await here. 

can you doc what's generally going on, and why we want to warm up caches per doucment.  is there a reason we do it at that level, as opposed to one cache per project?